addons/web_editor/static/src/js/editor/snippets.options.js

```diff
@@ -4454,10 +4463,154 @@ registry.ImageOptimize = ImageHandlerOption.extend({
+        svg.setAttribute('height', img.naturalHeight);
+        img.dataset.shape = shape;
+        await this._writeSvg(svg.outerHTML, fileName);
+        return this._applyColor(img.dataset.shapeColor, true, previewMode);
+    },
+
+    /**
+     * Applies the correct color to the shape by looking for the default palette's hexcode inside the OG svg files
+     * @param {string} color HEX Code for the color of the background shape
+     * @param {bool} newShape wether it's a new shape or not so we know to look for default colors or previous ones
+     * @param {string} previewMode state of the editor to know which colors to look for
+     */
+    async _applyColor(color, newShape, previewMode) {
+        const img = this._getImg();
+        if (!img.dataset.shape) {
+            return;
+        }
+
+        const fileName = img.dataset.shape.split('/')[1];
+        const clipShapeURL = img.src;
+        let clipShape = await (await fetch(clipShapeURL)).text();
+        let newColor = color && normalizeColor(color);
+        if (newShape) {
+            const DEFAULT_PALETTE = weUtils.DEFAULT_PALETTE;
+            const combinedColorsRegex = new RegExp(Object.values(DEFAULT_PALETTE).join('|'), 'gi');
+            if (previewMode === 'reset' && img.dataset.shapeColor) {
+                clipShape = clipShape.replace(combinedColorsRegex, img.dataset.shapeColor);
+            } else {
+                clipShape = clipShape.replace(combinedColorsRegex, match => {
+                    const colorId = Object.keys(DEFAULT_PALETTE).find(key => DEFAULT_PALETTE[key] === match.toUpperCase());
+                    return newColor = weUtils.getCSSVariableValue('o-color-' + colorId);
+                });
+            }
+        } else {
+            clipShape = clipShape.replace(img.dataset.shapeColor, newColor);
+        }
+        await this._writeSvg(clipShape, fileName);
+        if (!previewMode) {
+            img.dataset.shapeColor = newColor;
+        }
+    },
 
+    /**
+     * Replace the URL with a dataURL of the SVG converted in base64
+     * @param {string} svgText svg XML file as a string (inner content)
+     * @param {string} fileName name the file will take place as
+     * @returns returns a promise that's resolved once the svg is properly loaded in the document
+     */
+    async _writeSvg(svgText, fileName) {
+        const img = this._getImg();
+        const file = new File([svgText], fileName, {
+            type: 'image/svg+xml',
+        });
+
+        const reader = new FileReader();
+        const readPromise = new Promise(resolve => {
+            reader.addEventListener('load', () => resolve(reader.result));
+        });
+        reader.readAsDataURL(file);
+        const dataURL = await readPromise;
+        const wrotePromise = new Promise((resolve, reject) => {
+            img.addEventListener('load', () => resolve(), {once: true});
+            img.addEventListener('error', () => reject(), {once: true});
+        });
```

Comment:
Any reason we need to wait for the image to load?

Reply:
Not entierly sure if it can happen, but I did it to avoid race conditions where a user could trigger multiple events (e.g previewing a shape and instantly reset it, browser might trigger modifications on the previous shape). I thought that it might be a good idea to let the browser finish the loading of the image before triggering any other modifications. I am not sure however if it is needed as the data concerning the shape should already be set in `img.src`.